src/common/com/intellij/plugins/haxe/model/HaxeClassModel.java

```diff
@@ -0,0 +1,346 @@
+import com.intellij.plugins.haxe.lang.psi.*;
+import com.intellij.plugins.haxe.model.type.HaxeTypeResolver;
+import com.intellij.plugins.haxe.model.type.SpecificHaxeClassReference;
+import com.intellij.plugins.haxe.model.type.SpecificTypeReference;
+import com.intellij.plugins.haxe.util.HaxePsiUtils;
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiIdentifier;
+import org.apache.commons.lang.NotImplementedException;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.*;
+
+public class HaxeClassModel {
+  public HaxeClass haxeClass;
+
+  public HaxeClassModel(HaxeClass haxeClass) {
+    this.haxeClass = haxeClass;
+  }
+
+  public HaxeClassReferenceModel getParentClassReference() {
+    List<HaxeType> list = haxeClass.getHaxeExtendsList();
+    if (list.size() == 0) return null;
+    return new HaxeClassReferenceModel(list.get(0));
+  }
+
+  static public boolean isValidClassName(String name) {
+    return name.substring(0, 1).equals(name.substring(0, 1).toUpperCase());
+  }
+
+  public HaxeClassModel getParentClass() {
+    final HaxeClassReferenceModel reference = this.getParentClassReference();
+    return (reference != null) ? reference.getHaxeClass() : null;
+  }
+
+  public List<HaxeClassReferenceModel> getInterfaceExtendingInterfaces() {
+    List<HaxeType> list = haxeClass.getHaxeExtendsList();
+    List<HaxeClassReferenceModel> out = new ArrayList<HaxeClassReferenceModel>();
+    for (HaxeType type : list) {
+      out.add(new HaxeClassReferenceModel(type));
+    }
+    return out;
+  }
+
+  public List<HaxeClassReferenceModel> getImplementingInterfaces() {
+    List<HaxeType> list = haxeClass.getHaxeImplementsList();
+    List<HaxeClassReferenceModel> out = new ArrayList<HaxeClassReferenceModel>();
+    for (HaxeType type : list) {
+      out.add(new HaxeClassReferenceModel(type));
+    }
+    return out;
+  }
+
+  public boolean isExtern() {
+    return haxeClass.isExtern();
+  }
+
+  public boolean isClass() {
+    return !this.isAbstract() && (HaxeComponentType.typeOf(haxeClass) == HaxeComponentType.CLASS);
+  }
+
+  public boolean isInterface() {
+    return HaxeComponentType.typeOf(haxeClass) == HaxeComponentType.INTERFACE;
+  }
+
+  public boolean isEnum() {
+    return HaxeComponentType.typeOf(haxeClass) == HaxeComponentType.ENUM;
+  }
+
+  public boolean isTypedef() {
+    return HaxeComponentType.typeOf(haxeClass) == HaxeComponentType.TYPEDEF;
+  }
+
+  public boolean isAbstract() {
+    return haxeClass instanceof HaxeAbstractClassDeclaration;
+  }
+
+  // @TODO: Create AbstractHaxeClassModel extending this class for these methods?
+  // @TODO: this should be properly parsed in haxe.bnf so searching for the underlying type is not required
+  @Nullable
+  public HaxeTypeOrAnonymous getAbstractUnderlyingType() {
+    if (!isAbstract()) return null;
+    PsiElement[] children = getPsi().getChildren();
+    if (children.length >= 2) {
```

Comment:
PsiTreeUtil.getChildOfType is iterating over all the descendants? Of just its children?

Also this code fails when you do something like:
```
abstract Abstract<T>(OtherClass<T>) {
}
```

So this will require some more work.
I examined the AST and didn't find a proper way of doing that. I will have to check if checking for HaxeTypeOrAnonymous works, but probably not the best way of doing this. I think that changing the haxe.bnf for that part would be good.